### PR TITLE
Add initial performance metric codes and doc

### PR DESCRIPTION
**Pull request by ftessier**
I created a markdown log regarding the very early performance testing for basic codes that simply update arrays with random numbers, as discussed earlier. I also include all the source code files discussed in the document (they are also listed in the `.md` for the record), in case anyone wants to play with it. I am just short now of the interesting part, i.e., testing on the GPU; I need to complete a local CUDA installation first, might prove tricky on the work computer which I don't fully manage. @SimonBiggs feel free to merge this PR or leave it open for now, as you see fit.

### CODE REVIEW

**Review by SimonBiggs — APPROVED** [review]
Beautiful :slightly_smiling_face:

Of note, to get consistent in-code timings of JAX code I had to adjust the timer decorator as such:

```python
def _apply_blocking(function_return_item):
    try:
        function_return_item.block_until_ready()
        return
    except AttributeError:
        pass


def timer(f):
    def wrap(*args, **kwargs):
        start = time.time()
        ret = f(*args, **kwargs)

        # See https://jax.readthedocs.io/en/latest/async_dispatch.html
        # for why this is needed.
        for return_item in ret:
            _apply_blocking(return_item)

        stop = time.time()
        duration = (stop - start) * 1000.0
        print("run {}: {:.3f} ms".format(run, duration))
        return ret

    return wrap
```

And then, using the timer defined as above, on the CPU, I can replicate the 40 second initial loop:

```
--------------------------------------
Bare jax.numpy arrays: x = jax.numpy.zeros(NUM_PARTICLES)
--------------------------------------
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
run 0: 40563.615 ms
run 1: 819.536 ms
run 2: 899.708 ms
run 3: 812.876 ms
run 4: 812.641 ms
run 5: 819.078 ms
run 6: 841.647 ms
run 7: 825.635 ms
run 8: 822.212 ms
run 9: 815.643 ms
--------------------------------------
```

However, if I make one slight adjustment to the `runIterations` function to pre-calcing the random numbers instead, (not sure if that is a fair comparison... but it is something we should aim to pull out of inner loops when feasible):

```python
@jit
def runIterations(prng_key, x, y, z, u, v, w, E):
    random_values = jax.random.normal(prng_key, shape=(ITERATIONS, 7, NUM_PARTICLES))
    (prng_key,) = jax.random.split(prng_key, 1)

    for j in range(ITERATIONS):
        x += random_values[j, 0, :]
        y += random_values[j, 1, :]
        z += random_values[j, 2, :]
        u += random_values[j, 3, :]
        v += random_values[j, 4, :]
        w += random_values[j, 5, :]
        E += random_values[j, 6, :]

    return prng_key, x, y, z, u, v, w, E
```

I get the following timings:

```
--------------------------------------
Bare jax.numpy arrays: x = jax.numpy.zeros(NUM_PARTICLES)
--------------------------------------
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
run 0: 4920.727 ms
run 1: 689.873 ms
run 2: 685.726 ms
run 3: 687.475 ms
run 4: 688.189 ms
run 5: 688.421 ms
run 6: 690.956 ms
run 7: 686.045 ms
run 8: 785.624 ms
run 9: 789.801 ms
--------------------------------------
```

About a 15% improvement on the jitted code, and an order of magnitude improvement on the first loop.

File: performance/performance-metrics.md
Comment (SimonBiggs) [inline performance/performance-metrics.md]:
```suggestion
There is not much we can do to improve this simplistic C++ for performance (short of multithreading). We _could_ write a plain C version (or an assembler version for that matter, but we are not envisaging a plain C rewrite of EGSnrc. Hence, this constitutes our baseline performance (for 1e5 particles, 100 random updates, and 10 runs, on a 6-core 12-threads Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz, tallied with `perf --repeat 10`):
```

File: performance/performance-metrics.md
Comment (SimonBiggs) [inline performance/performance-metrics.md]:
```suggestion
Below is the `py-plain.py` and `perf` profile of the baseline source code, which relies on numpy (plain python is much slower, as expected). This implementation uses flat numpy arrays the default numpy random number generator, which is the same as in the C++ implementation, and one of the best generators available today (Mersenne Twister). We have verified that the timer decorator function does not incur a significant time cost.
```

**Review by SimonBiggs — COMMENTED** [review]
Just a couple of typos

**Comment by SimonBiggs** [comment]
@ftessier I believe you should have merge rights. Once you're happy with the PR feel free to merge :slightly_smiling_face:

**Comment by ftessier** [comment]
Thanks for noting the typos! 🙏  I had noted the change to the timer decorator to account the non-blocking jit calls, but if I understand correctly, the `task-clock:u` count captures the run time properly, while the original timer function provides a sense of the (non-blocked) code speed, _not counting_ the (non-blocked) jit calls. I though it might be useful to see both in the `perf` run.

**Pull request merged**
